zone/expedition_database.cpp

```diff
@@ -0,0 +1,730 @@
+		WHERE
+			character_id = {}
+			AND is_pending = TRUE
+			AND expedition_name = '{}';
+	), character_id, EscapeString(expedition_name));
+
+	database.QueryDatabase(query);
+}
+
+void ExpeditionDatabase::DeletePendingLockouts(uint32_t character_id)
+{
+	LogExpeditionsDetail("Deleting character [{}] pending lockouts", character_id);
+
+	auto query = fmt::format(SQL(
+		DELETE FROM character_expedition_lockouts
+		WHERE character_id = {} AND is_pending = TRUE;
+	), character_id);
+
+	database.QueryDatabase(query);
+}
+
+void ExpeditionDatabase::DeleteAllMembersPendingLockouts(const std::vector<ExpeditionMember>& members)
+{
+	LogExpeditionsDetail("Deleting pending lockouts for [{}] characters", members.size());
+
+	std::string query_character_ids;
+	for (const auto& member : members)
+	{
+		fmt::format_to(std::back_inserter(query_character_ids), "{},", member.char_id);
+	}
+
+	if (!query_character_ids.empty())
+	{
+		query_character_ids.pop_back(); // trailing comma
+
+		auto query = fmt::format(SQL(
+			DELETE FROM character_expedition_lockouts
+			WHERE character_id IN ({}) AND is_pending = TRUE;
+		), query_character_ids);
+
+		database.QueryDatabase(query);
+	}
+}
+
+void ExpeditionDatabase::DeleteLockout(uint32_t expedition_id, const std::string& event_name)
+{
+	LogExpeditionsDetail("Deleting expedition [{}] lockout event [{}]", expedition_id, event_name);
+
+	auto query = fmt::format(SQL(
+		DELETE FROM expedition_lockouts
+		WHERE expedition_id = {} AND event_name = '{}';
+	), expedition_id, EscapeString(event_name));
+
+	database.QueryDatabase(query);
+}
+
+uint32_t ExpeditionDatabase::GetExpeditionIDFromCharacterID(uint32_t character_id)
+{
+	LogExpeditionsDetail("Getting expedition id for character [{}]", character_id);
+
+	uint32_t expedition_id = 0;
+	auto query = fmt::format(SQL(
+		SELECT expedition_id FROM expedition_members WHERE character_id = {};
+	), character_id);
+
+	auto results = database.QueryDatabase(query);
+	if (results.Success() && results.RowCount() > 0)
+	{
+		auto row = results.begin();
+		expedition_id = strtoul(row[0], nullptr, 10);
+	}
+	return expedition_id;
+}
+
+void ExpeditionDatabase::InsertCharacterLockouts(
+	uint32_t character_id, const std::vector<ExpeditionLockoutTimer>& lockouts,
+	bool replace_timer, bool is_pending)
+{
+	LogExpeditionsDetail("Inserting [{}] lockouts for character [{}]", lockouts.size(), character_id);
+
+	std::string insert_values;
+	for (const auto& lockout : lockouts)
+	{
+		fmt::format_to(std::back_inserter(insert_values),
+			"({}, FROM_UNIXTIME({}), {}, '{}', '{}', '{}', {}),",
+			character_id,
+			lockout.GetExpireTime(),
+			lockout.GetDuration(),
+			lockout.GetExpeditionUUID(),
+			EscapeString(lockout.GetExpeditionName()),
+			EscapeString(lockout.GetEventName()),
+			is_pending
+		);
+	}
+
+	if (!insert_values.empty())
+	{
+		insert_values.pop_back(); // trailing comma
+
+		std::string on_duplicate;
+		if (replace_timer)
+		{
+			on_duplicate = SQL(
```

Comment:
The replace condition can optionally update an existing expiration time or not here, so didn't want to do a flat replace (and on a personal habit I always tend to avoid using replace into)